app/app_test.go

```diff
@@ -320,8 +320,16 @@ func assertServerHealth(ctx context.Context, t *testing.T, opts sdkclient.Option
 		t.Error(clientErr)
 	}
 
-	if _, err := c.CheckHealth(ctx, nil); err != nil {
-		t.Error(err)
+	// Give the server 1 second to become healthy.
```

Comment:
A client should not return from `Dial` unless the server is available. This is an important feature of our SDKs. The `Dial` call invokes `GetSystemInfo` eagerly to get information and check health. There is no reason that call should work but `CheckHealth` should not.

Reply:
You might want to refactor this whole thing to use the Go SDK's built-in [DevServer](https://pkg.go.dev/go.temporal.io/sdk@v1.25.1/testsuite#DevServer) with a specified path and use the dev server's client.

Reply:
> A client should not return from `Dial` unless the server is available and healthy. This is an important feature of our SDKs. The `Dial` call invokes `GetSystemInfo` eagerly to get information and check health. There is no reason that call should work but `CheckHealth` should not.

I see. We are often seeing failures like this one:

https://github.com/temporalio/cli/actions/runs/6317569213/job/17154638693#step:5:199

...and simply waiting a second makes them go away. I do see the GetSystemInfo call you mention:

- [`resp, err := wc.workflowService.GetSystemInfo(ctx, &workflowservice.GetSystemInfoRequest{})`](https://github.com/temporalio/sdk-go/blob/06c90fb3698708ef11d1992c8d69883d26a20c31/internal/internal_workflow_client.go#L1162)

But it seems like [`CheckHealth()`](https://github.com/temporalio/sdk-go/blob/06c90fb3698708ef11d1992c8d69883d26a20c31/internal/internal_workflow_client.go#L1118) does a gRPC health check, which is different:

- [`resp, err := healthpb.NewHealthClient(wc.conn).Check(ctx, &healthpb.HealthCheckRequest{ ...`](https://github.com/temporalio/sdk-go/blob/06c90fb3698708ef11d1992c8d69883d26a20c31/internal/internal_workflow_client.go#L1124)

I do not know why GetSystemInfo would succeed when the gRPC health check fails. I can dig into server internals to figure out why, but it seems the behavior is indeed different.

Reply:
> You might want to refactor this whole thing to use the Go SDK's built-in [DevServer](https://pkg.go.dev/go.temporal.io/sdk@v1.25.1/testsuite#DevServer) with a specified path and use the dev server's client.

That would defeat the whole point of this test, which is to launch the dev server using Temporal CLI.

Reply:
How so? The dev server launches Temporal CLI for you.

Reply:
> How so? The dev server launches Temporal CLI for you.

It does, but by default it doesn't actually test the version of the CLI in CI, because it downloads a released version of the CLI or pulls from the user's PATH. I want to test the version in the repo (possibly with whatever random developer changes are there). That's only possible by invoking the relevant CLI commands directly.

To test with the right version, I would have to make sure the CLI binary is actually built as part of running the tests, and this adds complexity to the tests and an additional layer of abstraction that I think is harmful—it obscures what the test actually does, because the arguments to `temporal server start-dev` are not immediately visible in the test, and we are basically inserting an external dependency in between the test and the thing we're testing. This is also likely to make the test itself more brittle (e.g. if something changes in go-sdk).

Regardless, none of that refactoring would actually help with these test failures, and such a change would be out-of-scope for this PR anyway. So I think we can probably continue this elsewhere.

Reply:
> I do not know why GetSystemInfo would succeed when the gRPC health check fails

Me neither, it's important that this is not the case or we have bigger problems with the SDK/server contract/expectations I think.

Reply:
@cretz How do you feel about:
1. Merging this as-is
2. Opening a separate issue to track down why the gRPC health check is failing even when GetSystemInfo returns successfully
3. Reverting the polling loop for the gRPC health check in this PR once (2) is resolved

Reply:
Sure, anything to fix the problem I'm ok with and I don't see why we need to block this issue if it's already broken.

Reply:
Opened https://github.com/temporalio/temporal/issues/5015

Reply:
> > How so? The dev server launches Temporal CLI for you.
> 
> It does, but by default it doesn't actually test the version of the CLI in CI, because it downloads a released version of the CLI or pulls from the user's PATH. I want to test the version in the repo (possibly with whatever random developer changes are there). That's only possible by invoking the relevant CLI commands directly.

You could do that with the test framework by giving it a path to an existing binary.

> To test with the right version, I would have to make sure the CLI binary is actually built as part of running the tests, and this adds complexity to the tests and an additional layer of abstraction that I think is harmful—it obscures what the test actually does, because the arguments to `temporal server start-dev` are not immediately visible in the test, and we are basically inserting an external dependency in between the test and the thing we're testing. This is also likely to make the test itself more brittle (e.g. if something changes in go-sdk).

I think we can discuss the pros and cons here.
One pro is that we don't have to rewrite logic that waits for the server to be "ready".

> Regardless, none of that refactoring would actually help with these test failures, and such a change would be out-of-scope for this PR anyway. So I think we can probably continue this elsewhere.

SGTM